Add runtime option

jociemoore opened the pull request:
Node 8.10 is losing support at the end of the year. 

My [previous PR](https://github.com/bam-lambda/bam/pull/181) updates the runtime configuration of newly created lambdas.

This PR adds a `runtime` option to the `redeploy` command so that users can update their existing lambdas to an actively supported runtime without downtime, and therefore, continue to use BAM! to maintain these functions.

BAM! is still more or less opinionated about the runtime configuration:
- You cannot configure the runtime when creating lambda.
- You cannot configure the runtime to a non-Node runtime when updating a lambda.
- You cannot configure the runtime to an Node runtime that has already reached EOL (or, in the case of Node 8.10, will soon reach its EOL).

jasonoverby commented on src/util/validationMessages.js:
```diff
@@ -36,6 +37,21 @@ const getCannotRemoveMethodMadeInConsoleMsg = (removeMethods, existingMethods) =
+  const totalValidRuntimes = validNodeRuntimes.length;
+  let validNodeRuntimesStr;
+
+  if (totalValidRuntimes === 2) {
```
How would `totalValidRuntimes` ever be something other than `1` when you have set `validNodeRuntimes` as an array with one element?

jociemoore replied:
@jasonoverby I implemented it this way in anticipation that AWS might one day support more than one node runtime. Updates to Node are [happening frequently enough](https://github.com/nodejs/Release) that related changes to our codebase could get annoying, so I just wanted to make easy as possible; with this in place, we just have to edit that array of `validNodeRuntimes` when there is an update. 

Now that I look at it again, it might nice to extract this type of information out and into a `settings.js` file.

jasonoverby replied:
Cool.  I thought you may have been doing that :)

jasonoverby commented:
I think this is a step in the right direction, but, though it would require additional work, it might be better to just prompt the user to update to `nodejs10.x` instead.  That could be done in a future PR.

jociemoore commented:
@jasonoverby I like your idea with prompting the user to update! I think it fits right in with this work and there really wasn't all that much more to do:
1) tweaked `redeploy.js` to validate the Node runtime every time
2) tweaked `validationMessages.js` to that the `invalidNodeRuntime` message would apply to both scenarios where a user doesn't know they need to update and where a user is trying to update the runtime but supplies an invalid cli arg. 

I also did the following cleanup:
- extracted `getLambdaConfig` functionality into its own file as I would have had to duplicate it
- updated the help command to include the new option
- merged in the [first PR](https://github.com/bam-lambda/bam/pull/181) in order to run the `updateLambda` test

Let me know if you have any other thoughts before I merge this work into develop and then into master!

jasonoverby commented on src/aws/getLambdaConfig.js:
```diff
@@ -0,0 +1,10 @@
+const { asyncGetFunction } = require('./awsFunctions');
```
💯

jasonoverby commented on src/util/validations.js:
```diff
@@ -95,6 +96,15 @@ const removeMethodsDeployedPreviouslyWithBam = async ({
+  return name.includes('node');
+};
+
+const runtimeIsValid = async (name) => {
```
This doesn't need to be `async`.

jasonoverby commented on src/util/validations.js:
```diff
@@ -95,6 +96,15 @@ const removeMethodsDeployedPreviouslyWithBam = async ({
   return result.length === 0;
 };
 
+const isANodeRuntime = async (name) => {
```
This doesn't need to be `async`.

jociemoore commented:
Actually, I think the `runtime` flag is necessary. Let me walk through the scenarios you suggested...

We could prompt the user to update the function's runtime when the runtime check fails during a  redeploy. While they could then go into the AWS console and make a change to the runtime configuration, they may not be able to make necessary changes to the code in the same "save"; a lambda may be too big and not editable via the UI. Changes to the code may be required if there are applicable breaking changes between Node versions. In this sense, there would be a time lag between when a user makes changes to the function, uploads it, and then changes the runtime in the console. Or visa versa. 

Likewise if we update the runtime for the user automatically; this would feel like surprising and potentially frustrating behavior.

@jasonoverby wdyt?

jasonoverby commented:
@jociemoore I was thinking that when a user redeploys we would go through these steps:

- check the Node version
  - if the Node version is `>= 10`+, allow the redeploy as usual
  - if the Node version is `< 10`
    - provide a message to the user letting them know that in order to redeploy, they must update the Node version
    - get `Y`/`N` input from the user regarding whether they want the Node version updated
      - if `Y`, `BAM!` updates the Node version as part of the redeploy
      - if `N`, `BAM!` lets the user know that the redeploy has been canceled and quits

This would eliminate the need for the flag, right?

jociemoore commented:
@jasonoverby I think the conversational route would work too. If there were more than one AWS-supported, Node runtimes, we would would have to think about how we would implement a selection list in the cli. 

Since there are only two months left until Node 8 reaches its EOL, I think it would be best to ship what we have currently and revisit the "flag or no flag" idea in a future PR. wdyt?

Note: I added to the runtime validation message so that a user is given a heads up about possible breaking changes.
